platform/Callback.h

```diff
@@ -35,209 +42,441 @@ namespace mbed {
+template <typename T>
+struct unqualify_fn { };
+
+// *INDENT-OFF*
+template <typename R, typename... Args>
+struct unqualify_fn<R(Args...)> : mstd::type_identity<R(Args...)> { };
+template <typename R, typename... Args>
+struct unqualify_fn<R(Args...) &> : mstd::type_identity<R(Args...)> { };
+template <typename R, typename... Args>
+struct unqualify_fn<R(Args...) const> : mstd::type_identity<R(Args...)> { };
+template <typename R, typename... Args>
+struct unqualify_fn<R(Args...) const &> : mstd::type_identity<R(Args...)> { };
+template <typename R, typename... Args>
+struct unqualify_fn<R(Args...) volatile> : mstd::type_identity<R(Args...)> { };
+template <typename R, typename... Args>
+struct unqualify_fn<R(Args...) volatile &> : mstd::type_identity<R(Args...)> { };
+template <typename R, typename... Args>
+struct unqualify_fn<R(Args...) const volatile> : mstd::type_identity<R(Args...)> { };
+template <typename R, typename... Args>
+struct unqualify_fn<R(Args...) const volatile &> : mstd::type_identity<R(Args...)> { };
+#if __cplusplus >=201703 || __cpp_noexcept_function_type >= 201510
+/* We have to spell out all c/v/ref/noexcept versions here, as specialization needs exact type match */
+/* Compare to callback() and deduction guides, where dropping the noexcept is a permitted conversion */
+template <typename R, typename... Args>
+struct unqualify_fn<R(Args...) noexcept> : mstd::type_identity<R(Args...)> { };
+template <typename R, typename... Args>
+struct unqualify_fn<R(Args...) & noexcept> : mstd::type_identity<R(Args...)> { };
+template <typename R, typename... Args>
+struct unqualify_fn<R(Args...) const noexcept> : mstd::type_identity<R(Args...)> { };
+template <typename R, typename... Args>
+struct unqualify_fn<R(Args...) const & noexcept> : mstd::type_identity<R(Args...)> { };
+template <typename R, typename... Args>
+struct unqualify_fn<R(Args...) volatile noexcept> : mstd::type_identity<R(Args...)> { };
+template <typename R, typename... Args>
+struct unqualify_fn<R(Args...) volatile & noexcept> : mstd::type_identity<R(Args...)> { };
+template <typename R, typename... Args>
+struct unqualify_fn<R(Args...) const volatile noexcept> : mstd::type_identity<R(Args...)> { };
+template <typename R, typename... Args>
+struct unqualify_fn<R(Args...) const volatile & noexcept> : mstd::type_identity<R(Args...)> { };
+#endif
+
+template <typename T>
+using unqualify_fn_t = typename unqualify_fn<T>::type;
+
+template <typename R, typename F, typename... Args, typename std::enable_if_t<!std::is_void<R>::value, int> = 0>
+R invoke_r(F&& f, Args&&... args)
+{
+    return mstd::invoke(std::forward<F>(f), std::forward<Args>(args)...);
+}
+
+template <typename R, typename F, typename... Args, typename std::enable_if_t<std::is_void<R>::value, int> = 0>
+R invoke_r(F&& f, Args&&... args)
+{
+    static_cast<void>(mstd::invoke(std::forward<F>(f), std::forward<Args>(args)...));
```

Comment:
Hmm. That was lifted from the wording of the standard

> Define `INVOKE<R>(f, t1, t2, . . . , tN)` as `static_cast<void>(INVOKE(f, t1, t2, . . . , tN))` if R is
cv void, otherwise INVOKE(f, t1, t2, . . . , tN) implicitly converted to R.

I guess that's just to make a void-type expression though - you can't "implicitly convert" to void, so you need to cast it. I think I thought something more subtle was happening, but the blurb about static_cast just says:

> Any expression can be explicitly converted to type cv void, in which case it becomes a discarded-value expression (7.2).

So yes, I believe we can drop the cast.